In [28]:
import Bio.AlignIO as AlignIO

- `parse` vrne iterator MultipleSeqAlignement iterator (več sekvenc)
- `read` vrne iterator za eno sekvenco, idk..

In [8]:
alignments = AlignIO.read("PF02171_seed.sth", "stockholm")
alignments

<<class 'Bio.Align.MultipleSeqAlignment'> instance (16 records of length 395) at 7f6f8c41e510>

Iterator instance with 16 records (SeqRecord) of length 395.

Za vsak zapis izpišite opis, začetno in končno točko, zaporedje in zunanje db_xrefs (če obstajajo).j

In [15]:
for alignment in alignments:
    opis = alignment.description
    start = alignment.annotations["start"]
    end = alignment.annotations["end"]
    zaporedje = alignment.seq
    db_xrefs = "".join([x for x in alignment.dbxrefs])
    db_xrefs = "None" if db_xrefs == "" else db_xrefs

    print(f"""
    opis: {opis}
    start: {start}
    end: {end}
    zaporedje: {zaporedje}
    db_xrefs: {db_xrefs}
    """)


    opis: AGO1_SCHPO/500-799
    start: 500
    end: 799
    zaporedje: YLFFILDK-NSPEP-YGSIKRVCNTMLGVPSQCAISKHILQS---------KPQYCANLGMKINVKVGGIN-CSLIPKSNP----LGNVPTL---------ILGGDVYHPGVGA----------TGVSIASIVASVD-LNGCKYTAVSRSQPRHQEVIEG-MKD------------IVVYLLQGFRAMTKQ-QPQRIIYFRDGTSEGQFLSVINDELSQIKEACH-------SLSPKYN--PKILVCTTQKRHHARFFIKNKSDG----------------------DRNGNPLPGTII---EKHVTHPYQYDFYLISHPSLQGVSVPVHYTVLHDEIQMPPDQF-QTL------CYNLCYVYARAT----SAVSLVPPVYYAHLVSNLARYQDV
    db_xrefs: None
    

    opis: AGO6_ARATH/541-851
    start: 541
    end: 851
    zaporedje: FILCILPERKTSDI-YGPWKKICLTEEGIHTQCICPIKI------------SDQYLTNVLLKINSKLGGIN-SLLGIEYSYNIPLINKIPTL---------ILGMDVSHGPPGR---------ADVPSVAAVVGSKCWPLISRYRAAVRTQSPRLEMIDSLFQPIENTE--KGDNGIMNELFVEFYRTSRARKPKQIIIFRDGVSESQFEQVLKIEVDQIIKAYQ-------RLGESDV--PKFTVIVAQKNHHTKLFQAKGPE---------------------------NVPAGTVV---DTKIVHPTNYDFYMCAHAGKIGTSRPAHYHVLLDEIGFSPDDL-QNL------IHSLSYVNQRST----TATSIVAPVRYAHLAAAQVAQFTK
    db_xrefs: None
    

    opis: AGO

Ustvarimo nekaj poravnav več zaporedij in jih shrnaimo v formatu phylip. Izhodna datoteka naj bo my_malign.phy. Na koncu preberite in izpišite vsebino datoteke.

In [17]:
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio.Align import MultipleSeqAlignment

In [18]:
align1 = MultipleSeqAlignment([
    SeqRecord(Seq("ACTGCTAGCTAG"), id=" Alpha"),
    SeqRecord(Seq("ACT-CTAGCTAG"), id=" Beta"),
    SeqRecord(Seq("ACTGCTAGDTAG"), id=" Gamma")
])

align2 = MultipleSeqAlignment([
    SeqRecord (Seq("GTCAGC-AG"), id="Delta"),
    SeqRecord (Seq("GACAGCTAG"), id="Epsilon"),
    SeqRecord (Seq("GTCAGCTAG"), id="Zeta")
])

align3 = MultipleSeqAlignment([
    SeqRecord(Seq("ACTAGTACAGCTG"), id="Eta"),
    SeqRecord(Seq("ACTAGTACAGCT-"), id="Theta"),
    SeqRecord(Seq("-CTACTACAGGTG"), id="lota")
])

In [20]:
my_alignements = [align1, align2, align3]
N = AlignIO.write(my_alignements, "my_malign.phy", "phylip") # vrne število vnosov

In [21]:
N

3

In [22]:
for alignment in AlignIO.parse("my_malign.phy", "phylip"):
    print(alignment)

Alignment with 3 rows and 12 columns
ACTGCTAGCTAG Alpha
ACT-CTAGCTAG Beta
ACTGCTAGDTAG Gamma
Alignment with 3 rows and 9 columns
GTCAGC-AG Delta
GACAGCTAG Epsilon
GTCAGCTAG Zeta
Alignment with 3 rows and 13 columns
ACTAGTACAGCTG Eta
ACTAGTACAGCT- Theta
-CTACTACAGGTG lota


Lahko pretvarjamo med formati, če so vsi potrebni podatki dostopni.

Pretvori stockholmski format v clustal.

In [24]:
AlignIO.convert("PF02171_seed.sth", "stockholm", "PF02171.aln", "clustal")

1

In [27]:
import os
os.listdir()

['PF02171_seed.sth',
 'v6.ipynb',
 'E06_Poravnava_zaporedij.pptx',
 'my_malign.phy',
 'PF02171.aln']

Za vsak zapis datoteke PF02171_seed.sth izpiši opis, prvih 20 baz in zadnjih 20 baz.

In [42]:
alignments = AlignIO.read("PF02171_seed.sth", "stockholm")
wanted_alignments = alignments[:, :20] + alignments[:, -20:]

In [43]:
for alignment in wanted_alignments:
    print(alignment.description, "\t", alignment.seq)

AGO1_SCHPO/500-799 	 YLFFILDK-NSPEP-YGSIKLVPPVYYAHLVSNLARYQDV
AGO6_ARATH/541-851 	 FILCILPERKTSDI-YGPWKIVAPVRYAHLAAAQVAQFTK
AGO4_ARATH/577-885 	 FILCVLPDKKNSDL-YGPWKVVAPICYAHLAAAQLGTFMK
TAG76_CAEEL/660-966 	 CIIVVLQS-KNSDI-YMTVKIPTPVYYADLVATRARCHVK
O16720_CAEEL/566-867 	 LIVVVLPG--KTPI-YAEVKIPAPAYYAHLVAFRARYHLV
O62275_CAEEL/594-924 	 TFVFIITD-DSITT-LHQRYLPTPLYVANEYAKRGRNLWN
YQ53_CAEEL/650-977 	 DILVGIAR-EKKPD-VHDILVPDVLYAAENLAKRGRNNYK
NRDE3_CAEEL/673-1001 	 TIVFGIIA-EKRPD-MHDILIPNVSYAAQNLAKRGHNNYK
Q17567_CAEEL/397-708 	 MLVVMLAD-DNKTR-YDSLKVPAPCQYAHKLAFLTAQSLH
AUB_DROME/555-852 	 IVMVVMRS-PNEEK-YSCIKVPAVCHYAHKLAFLVAESIN
PIWI_DROME/538-829 	 LILCLVPN-DNAER-YSSIKVPAVCQYAKKLATLVGTNLH
PIWL1_HUMAN/555-847 	 IVVCLLSS-NRKDK-YDAIKVPAPCQYAHKLAFLVGQSIH
PIWI_ARCFU/110-406 	 GIMLVLPE-YNTPL-YYKLKLPVTVNYPKLVAGIIANVNR
Y1321_METJA/426-699 	 CFALIIGKEKYKDNDYYEILIPAPIHYADKFVKALGKNWK
O67434_AQUAE/419-694 	 LVIVFLEEYPKVDP-YKSFLLPATVHYSDKITKLMLRGIE
AGO10_ARATH/625-946 	 LLLAILPD-NNGSL-YGDLKIVPPAYYAHLAAFRAR

Za pairwise aligment uporabimo `PairwiseAligner` za globalno ali lokalno poravnavo.

In [44]:
from Bio.Align import PairwiseAligner

In [46]:
aligner = PairwiseAligner(match_score=1) # dobesedno to kar piše

Uporabimo aligner za izračun ocene poravnave dveh sekvenc.

In [47]:
target = "GAACT"
query = "GAT"
score = aligner.score(target, query)
score

3.0

`PairwiseAligner.align` vrne vse možne poravnave nad dvema sekvencama. Vsak `Alignment` objekt hrani `score`, `target`, `query`...

Poravnava je shranjena v obliki koordinat zaporedja (`coordinates`).

Dolžina poravnave je shranjena v `length` spremenljivki.

In [53]:
for a in aligner.align(target, query):
    print(f"""
    target: {a.target}
    query: {a.query}
    score: {a.score}
    length: {a.length}
    """)


    target: GAACT
    query: GAT
    score: 3.0
    length: 5
    

    target: GAACT
    query: GAT
    score: 3.0
    length: 5
    


Poravnavi dobita enak score, če se poravnava razlikuje le v položaju vrzeli (gap).

In [63]:
aligner.mismatch_score = -10
alignments = aligner.align("AACAA", "AAGAA")

In [69]:
for a in alignments:
    print(a, end="")
    print("score: ", a.score)

target            0 AAC-AA 5
                  0 ||--|| 6
query             0 AA-GAA 5
score:  4.0
target            0 AA-CAA 5
                  0 ||--|| 6
query             0 AAG-AA 5
score:  4.0


Privzeto se uporavlja globalna poravnava, ki najde optimalno poravnavo po celotnih zaporedjih. Z lokalno poravnano se najde podzaporedje z najvišjo oceno poravnave.

In [73]:
aligner.mode = "global"

for a  in aligner.align("AGAACTC", "GAACT"):
    print(a)

target            0 AGAACTC 7
                  0 -|||||- 7
query             0 -GAACT- 5



In [74]:
aligner.mode = "local"

for a  in aligner.align("AGAACTC", "GAACT"):
    print(a)

target            1 GAACT 6
                  0 ||||| 5
query             0 GAACT 5



Če izpišemo objekt `PairwiseAligner` se izpišejo vsi parametri, ki jih trenutno uporablja.

`PairwiseAligner.epsilon` je vrednost, ki se uporablja za mejo natančnosti poravnave (tako da je poravnava statistično značilna). Privzeto je enaka $10^{-6}$. Dve poravnavi se štejeta za enaki, če je absolutna razlika njunih ocen manjša od $\epsilon$.

In [76]:
aligner

'Smith-Waterman'

In [77]:
seq1 = Seq("MVLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPH"
           "FDLSHGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKL"
           "RVDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKYR")

seq2 = Seq("MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFESFG"
           "DLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHC"
           "DKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH")

In [80]:
aligner.mode = "global"
aligner.score(seq1, seq2)

72.0

In [82]:
# aligner.align(seq1, seq2)

OverflowError: number of optimal alignments is larger than 9223372036854775807

Poravnav je preveč in dobimo OverflowError.